File: packages/telemetry/openclaw/src/messages.ts
/**
 * Normalizes provider-specific message shapes into the parts-based GenAI
 * format Latitude's parser expects:
 *
 *   { role: "system" | "user" | "assistant" | "tool", parts: MessagePart[] }
 *
 * Dialects handled, dispatched per item on the item's own shape:
 *
 *   - OpenClaw / pi-ai transcript: `{role:"user", content: string | (text|image)[]}`,
 *     `{role:"assistant", content: (text|thinking|toolCall)[]}`,
 *     `{role:"toolResult", toolCallId, toolName, content, isError}`,
 *     `{role:"custom", ...}` runtime notes.
 *   - Anthropic: `{role, content: string | ContentBlock[]}` with `text`,
 *     `tool_use`, `tool_result`, `image`, `thinking` blocks.
 *   - OpenAI Chat Completions: `{role, content: string, tool_calls?}` and
 *     `{role:"tool", tool_call_id, content}`.
 *   - Already-normalized parts-shape messages, passed through unchanged.
 *
 * Anything unrecognized becomes a JSON-stringified text part so nothing is
 * silently dropped.
 */

export interface MessagePart {
  type: string
  content?: string
  id?: string | undefined
  name?: string | undefined
  arguments?: unknown
  response?: unknown
  modality?: string
  uri?: string
  [key: string]: unknown
}

export type MessageRole = "system" | "user" | "assistant" | "tool"

export interface Message {
  role: MessageRole
  parts: MessagePart[]
}

const ALLOWED_ROLES: ReadonlySet<MessageRole> = new Set(["system", "user", "assistant", "tool"])

export function normalizeMessage(raw: unknown): Message | undefined {
  if (!raw || typeof raw !== "object") return undefined
  const obj = raw as Record<string, unknown>

  if (Array.isArray(obj.parts)) {
    const parts: MessagePart[] = []
    for (const p of obj.parts) {
      if (p && typeof p === "object") parts.push(p as MessagePart)
    }
    return { role: coerceRole(obj.role), parts: parts.length > 0 ? parts : [{ type: "text", content: safeJson(raw) }] }
  }

  if (obj.role === "toolResult") return normalizeToolResult(obj)
  if (obj.role === "custom") return normalizeCustom(obj)

  const content = obj.content ?? obj.text ?? obj.message
  const role = isRuntimeContext(obj, content) ? "system" : coerceRole(obj.role)

  if (role === "tool" && obj.tool_call_id !== undefined) {
    return {
      role,
      parts: [
        {
          type: "tool_call_response",
          id: typeof obj.tool_call_id === "string" ? obj.tool_call_id : "",
          response: content ?? safeJson(obj),
        },
      ],
    }
  }

  if (typeof content === "string") {
    const parts: MessagePart[] = []
    if (content.length > 0) parts.push({ type: "text", content })
    appendToolCalls(parts, obj.tool_calls)
    if (parts.length === 0) parts.push({ type: "text", content: "" })
    return { role, parts }
  }

  if (Array.isArray(content)) {
    const parts: MessagePart[] = []
    for (const block of content) {
      const part = normalizeBlock(block)
      if (part) parts.push(part)
    }
    appendToolCalls(parts, obj.tool_calls)
    if (parts.length === 0) parts.push({ type: "text", content: safeJson(content) })
    return { role, parts }
  }

  return { role, parts: [{ type: "text", content: safeJson(raw) }] }
}

export function normalizeMessages(raw: readonly unknown[]): Message[] {
  const out: Message[] = []
  for (const m of raw) {
    const norm = normalizeMessage(m)
    if (norm) out.push(norm)
  }
  return out
}

export function userMessageFromPrompt(prompt: string): Message {
  return { role: "user", parts: [{ type: "text", content: prompt }] }
}

export function assistantMessageFromOutput(assistantTexts: readonly string[], lastAssistant: unknown): Message {
  if (lastAssistant !== undefined) {
    const norm = normalizeMessage(lastAssistant)
    if (norm) return { ...norm, role: "assistant" }
  }
  const parts: MessagePart[] = []
  for (const text of assistantTexts) {
    if (text.length > 0) parts.push({ type: "text", content: text })
  }
  if (parts.length === 0) parts.push({ type: "text", content: "" })
  return { role: "assistant", parts }
}

export function systemInstructionsParts(prompt: string): MessagePart[] {
  return [{ type: "text", content: prompt }]
}

// ─── Helpers ────────────────────────────────────────────────────────────────

const RUNTIME_CONTEXT_PREFIX = "[openclaw.runtime-context]"

/**
 * OpenClaw injects its per-turn context as an extra user message right before
 * the prompt. It is instructions rather than the user's words, so it renders
 * as a system message instead of a second user bubble.
 */
function isRuntimeContext(obj: Record<string, unknown>, content: unknown): boolean {
  if (obj.role !== "user") return false
  if (obj.runtimeContext) return true
  const text = typeof content === "string" ? content : Array.isArray(content) ? blockText(content[0]) : undefined
  return typeof text === "string" && text.trimStart().startsWith(RUNTIME_CONTEXT_PREFIX)
}

function coerceRole(raw: unknown): MessageRole {
  if (typeof raw !== "string") return "user"
  return ALLOWED_ROLES.has(raw as MessageRole) ? (raw as MessageRole) : "user"
}

function normalizeToolResult(obj: Record<string, unknown>): Message {
  const content = obj.content
  let response: unknown
  if (typeof content === "string") response = content
  else if (Array.isArray(content)) {
    const texts: string[] = []
    let hasNonText = false
    for (const block of content) {
      const text = blockText(block)
      if (text !== undefined) texts.push(text)
      else hasNonText = true
    }
    response = hasNonText ? content.map(normalizeBlock).filter(Boolean) : texts.join("\n")
  } else response = content ?? ""
  return {
    role: "tool",
    parts: [
      {
        type: "tool_call_response",
        id: typeof obj.toolCallId === "string" ? obj.toolCallId : "",
        name: typeof obj.toolName === "string" ? obj.toolName : undefined,
        response,
        ...(obj.isError === true ? { is_error: true } : {}),
      },
    ],
  }
}

/** Text of a plain text block, or of the `toolResult` block the Codex harness wraps tool output in. */
function blockText(block: unknown): string | undefined {
  if (!block || typeof block !== "object") return undefined
  const b = block as Record<string, unknown>
  if (b.type === "text" && typeof b.text === "string") return b.text
  if (b.type === "toolResult") {
    if (typeof b.text === "string") return b.text
    if (typeof b.content === "string") return b.content
  }
  return undefined
}

function normalizeCustom(obj: Record<string, unknown>): Message {
  const text = typeof obj.content === "string" ? obj.content : safeJson(obj)
  const customType = typeof obj.customType === "string" ? obj.customType : "custom"
  return { role: "user", parts: [{ type: "text", content: `[${customType}] ${text}` }] }
}

function normalizeBlock(raw: unknown): MessagePart | undefined {
  if (typeof raw === "string") return { type: "text", content: raw }
  if (!raw || typeof raw !== "object") return undefined
  const obj = raw as Record<string, unknown>
  const type = typeof obj.type === "string" ? obj.type : "text"

  if (type === "text") {
    if (typeof obj.content === "string") return { type: "text", content: obj.content }
    if (typeof obj.text === "string") return { type: "text", content: obj.text }
  }
  if (type === "toolResult") {
    const text = blockText(obj)
    return { type: "text", content: text ?? safeJson(raw) }
  }
  if (type === "toolCall" || type === "tool_use" || type === "tool_call") {
    return {
      type: "tool_call",
      id: typeof obj.id === "string" ? obj.id : "",
      name: typeof obj.name === "string" ? obj.name : "",
      arguments: obj.arguments ?? obj.input ?? {},
    }
  }
  if (type === "tool_result") {
    return {
      type: "tool_call_response",
      id: typeof obj.tool_use_id === "string" ? obj.tool_use_id : "",
      response: obj.content ?? "",
    }
  }
  if (type === "tool_call_response") {
    return {
      type: "tool_call_response",
      id: typeof obj.id === "string" ? obj.id : "",
      response: obj.response ?? "",
    }
  }
  if (type === "thinking") {
    if (typeof obj.thinking === "string" && obj.thinking.length > 0) return { type: "reasoning", content: obj.thinking }
    if (obj.redacted === true) return { type: "reasoning", content: "[redacted]" }
    return undefined
  }
  if (type === "reasoning" && typeof obj.content === "string") {
    return { type: "reasoning", content: obj.content }
  }
  if (type === "image") {
    const uri = imageUri(obj)
    if (uri) return { type: "uri", modality: "image", uri }
    return { type: "text", content: "[image]" }
  }

  return { type, content: safeJson(raw) }
}

function imageUri(obj: Record<string, unknown>): string | undefined {
  if (typeof obj.data === "string" && obj.data.length > 0) {
    const mime = typeof obj.mimeType === "string" ? obj.mimeType : "image/unknown"
    return `data:${mime};base64,${obj.data}`
  }
  if (obj.source && typeof obj.source === "object") {
    const src = obj.source as { media_type?: string; data?: string; url?: string }
    if (src.url) return src.url
    if (src.data) return `data:${src.media_type ?? "image/unknown"};base64,${src.data}`
  }
  return undefined
}

function appendToolCalls(parts: MessagePart[], raw: unknown): void {
  if (!Array.isArray(raw)) return
  for (const tc of raw) {
    if (!tc || typeof tc !== "object") continue
    const t = tc as Record<string, unknown>
    const fn = t.function as { name?: string; arguments?: string | Record<string, unknown> } | undefined
    let parsedArgs: unknown = fn?.arguments
    if (typeof parsedArgs === "string") {
      try {
        parsedArgs = JSON.parse(parsedArgs)
      } catch {
        // leave as string
      }
    }
    parts.push({
      type: "tool_call",
      id: typeof t.id === "string" ? t.id : "",
      name: fn?.name ?? "",
      arguments: parsedArgs ?? {},
    })
  }
}

function safeJson(value: unknown): string {
  try {
    if (typeof value === "string") return value
    return JSON.stringify(value)
  } catch {
    return ""
  }
}

File: packages/telemetry/prime-intellect/src/latitude_telemetry_prime_intellect/messages.py
"""Normalize Verifiers / OpenAI-shaped messages into Latitude GenAI parts."""

from __future__ import annotations

import json
from typing import Any, Dict, List, Optional

from .util import _get, _safe_json

_ROLES = {"system", "user", "assistant", "tool"}


def _normalize_messages(raw: Any) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    if not isinstance(raw, (list, tuple)):
        return out
    for m in raw:
        n = _normalize_message(m)
        if n:
            out.append(n)
    return out


def _normalize_message(m: Any) -> Optional[Dict[str, Any]]:
    if m is None:
        return None
    if isinstance(m, dict):
        role = m.get("role")
        role = role if role in _ROLES else "user"
        if role == "tool":
            return {
                "role": "tool",
                "parts": [
                    {
                        "type": "tool_call_response",
                        "id": m.get("tool_call_id") or "",
                        "response": _tool_result(m.get("content")),
                    }
                ],
            }
        return _content_message(role, m.get("content"), m, m.get("reasoning_content"))

    role = _get(m, "role")
    role = role if role in _ROLES else "user"
    if role == "tool":
        return {
            "role": "tool",
            "parts": [
                {
                    "type": "tool_call_response",
                    "id": _get(m, "tool_call_id") or "",
                    "response": _tool_result(_get(m, "content")),
                }
            ],
        }
    envelope = {
        "tool_calls": _get(m, "tool_calls"),
    }
    return _content_message(role, _get(m, "content"), envelope, _get(m, "reasoning_content"))


def _content_message(
    role: str,
    content: Any,
    envelope: Dict[str, Any],
    reasoning: Any = None,
) -> Optional[Dict[str, Any]]:
    parts: List[Dict[str, Any]] = []
    if isinstance(reasoning, str) and reasoning.strip():
        parts.append({"type": "reasoning", "content": reasoning})
    if isinstance(content, str):
        if content.strip():
            parts.append({"type": "text", "content": content})
    elif isinstance(content, list):
        for block in content:
            p = _block(block)
            if p:
                parts.append(p)
    elif content is not None:
        parts.append({"type": "text", "content": _safe_json(content)})
    _append_tool_calls(parts, envelope.get("tool_calls") if isinstance(envelope, dict) else None)
    if not parts:
        return None
    return {"role": role, "parts": parts}


def _block(block: Any) -> Optional[Dict[str, Any]]:
    if isinstance(block, str):
        if not block.strip():
            return None
        return {"type": "text", "content": block}
    if not isinstance(block, dict):
        return None
    btype = block.get("type") or "text"
    if btype == "text":
        text = block.get("content") or block.get("text") or ""
        if not isinstance(text, str) or not text.strip():
            return None
        return {"type": "text", "content": text}
    if btype in ("thinking", "reasoning"):
        text = block.get("thinking") or block.get("content") or ""
        if not isinstance(text, str) or not text.strip():
            return None
        return {"type": "reasoning", "content": text}
    return {"type": btype, "content": _safe_json(block)}


def _append_tool_calls(parts: List[Dict[str, Any]], raw: Any) -> None:
    if not isinstance(raw, (list, tuple)):
        return
    for tc in raw:
        fn = _get(tc, "function")
        name = _get(fn, "name") if fn is not None else _get(tc, "name")
        args = _get(fn, "arguments") if fn is not None else None
        if args is None:
            args = _get(tc, "arguments")
        if isinstance(args, str):
            try:
                args = json.loads(args)
            except Exception:
                pass
        parts.append(
            {
                "type": "tool_call",
                "id": _get(tc, "id") or "",
                "name": name or "",
                "arguments": args if args is not None else {},
            }
        )


def _tool_result(raw: Any) -> Any:
    if isinstance(raw, str):
        return raw
    if isinstance(raw, list):
        texts = [b.get("text") if isinstance(b, dict) else b for b in raw]
        joined = "\n".join(t for t in texts if isinstance(t, str))
        return joined or raw
    return raw


def _message_role(m: Any) -> str:
    role = _get(m, "role") if not isinstance(m, dict) else m.get("role")
    return role if isinstance(role, str) else ""


def _system_prompt_from_messages(messages: Any) -> Optional[str]:
    if not isinstance(messages, (list, tuple)):
        return None
    for m in messages:
        if _message_role(m) != "system":
            continue
        content = _get(m, "content") if not isinstance(m, dict) else m.get("content")
        if isinstance(content, str) and content.strip():
            return content
    return None


def _first_user_text(messages: Any) -> Optional[str]:
    if not isinstance(messages, (list, tuple)):
        return None
    for m in messages:
        if _message_role(m) != "user":
            continue
        content = _get(m, "content") if not isinstance(m, dict) else m.get("content")
        if isinstance(content, str) and content.strip():
            return content
    return None
